src/ffi/mod.rs

```diff
@@ -29,182 +30,93 @@
-pub use self::eval::*;
-pub use self::fileobject::*;
-pub use self::fileutils::*;
-pub use self::floatobject::*;
 #[cfg(not(Py_LIMITED_API))]
-pub use self::funcobject::*;
-#[cfg(not(Py_LIMITED_API))]
-pub use self::genobject::*;
-pub use self::import::*;
-pub use self::intrcheck::*;
-pub use self::iterobject::*;
-pub use self::listobject::*;
-pub use self::longobject::*;
-pub use self::marshal::*;
-pub use self::memoryobject::*;
-pub use self::methodobject::*;
-pub use self::modsupport::*;
-pub use self::moduleobject::*;
-pub use self::object::*;
-pub use self::objimpl::*;
-pub use self::osmodule::*;
+mod cpython;
 #[cfg(not(Py_LIMITED_API))]
-pub use self::pyarena::*;
-pub use self::pycapsule::*;
-pub use self::pyerrors::*;
-pub use self::pyframe::*;
-pub use self::pyhash::*;
-pub use self::pylifecycle::*;
-pub use self::pymem::*;
-pub use self::pyport::*;
-pub use self::pystate::*;
-pub use self::pystrtod::*;
-pub use self::pythonrun::*;
-pub use self::rangeobject::*;
-pub use self::setobject::*;
-pub use self::sliceobject::*;
-pub use self::structseq::*;
-pub use self::sysmodule::*;
-pub use self::traceback::*;
-pub use self::tupleobject::*;
-pub use self::typeslots::*;
-pub use self::unicodeobject::*;
-pub use self::warnings::*;
-pub use self::weakrefobject::*;
-
-mod abstract_;
-// skipped asdl.h
-// skipped ast.h
-mod bltinmodule;
-mod boolobject;
-mod bytearrayobject;
-mod bytesobject;
-// skipped cellobject.h
-mod ceval;
-// skipped classobject.h
-mod code;
-mod codecs;
-mod compile;
-mod complexobject;
-#[cfg(all(Py_3_8, not(Py_LIMITED_API)))]
-mod context; // It's actually 3.7.1, but no cfg for patches.
+mod datetime;
+
 #[cfg(not(Py_LIMITED_API))]
-pub(crate) mod datetime;
-mod descrobject;
-mod dictobject;
-// skipped dynamic_annotations.h
-mod enumobject;
-// skipped errcode.h
-mod eval;
-// skipped exports.h
-mod fileobject;
-mod fileutils;
-mod floatobject;
-// skipped empty frameobject.h
+pub use datetime::*;
+
 #[cfg(not(Py_LIMITED_API))]
-pub(crate) mod funcobject;
-// skipped genericaliasobject.h
+use {libc::c_int, std::ops::Deref};
+
+//  reexport raw bindings exposed in pyo3_ffi
+pub use pyo3_ffi::*;
+
+/// Helper to enable #\[pymethods\] to see the workaround for __ipow__ on Python 3.7
+#[doc(hidden)]
+pub use crate::impl_::pymethods::ipowfunc;
+
+// Shadow raw ffi bindings for backwards compatability
+
 #[cfg(not(Py_LIMITED_API))]
-mod genobject;
-mod import;
-// skipped interpreteridobject.h
-mod intrcheck;
-mod iterobject;
-mod listobject;
-// skipped longintrepr.h
-mod longobject;
-pub(crate) mod marshal;
-mod memoryobject;
-mod methodobject;
-mod modsupport;
-mod moduleobject;
-// skipped namespaceobject.h
-mod object;
-mod objimpl;
-// skipped odictobject.h
-// skipped opcode.h
-// skipped osdefs.h
-mod osmodule;
-// skipped parser_interface.h
-// skipped patchlevel.h
-// skipped picklebufobject.h
-// skipped pyctype.h
-// skipped py_curses.h
+pub static PyDateTimeAPI: _PyDateTimeAPI_impl = _PyDateTimeAPI_impl {};
```

Comment:
I actually think we should remove these `PyDateTimeAPI` and `PyDateTime_TimeZone_UTC` statics, I don't think it's good for them to differ from the ffi crate.

Reply:
I have removed these statics and adjusted the save code. This actually simplified the code as all access to these statics (and also the appropriate functions) only happens when the GIL is already acquired which removed the need to call `Python::aquire_gil`. While this is trivial to see in most places (as a `Python` instance is already available) I had to use `Python::assume_gil` in the  `pyobject_native_type!` macros as these don't provide direct access to a `Python` instance. However I am sure this is correct from looking at the source code. Nonetheless I marked all instances of this with a `REVIEW` comment. It would be good to check that my understand is correct here